rootfs/etc/nginx/lua/balancer/ewma.lua

```diff
@@ -22,28 +22,50 @@ local function decay_ewma(ewma, last_touched_at, rtt, now)
   return ewma
 end
 
-local function get_or_update_ewma(self, upstream, rtt, update)
-  local ewma = self.ewma[upstream] or 0
+local function get_or_update_ewma(upstream, rtt, update)
+  -- When not using a lock, it is possible that an another worker updates ewma
+  -- value of upstream after we get the value but before we update and store it back.
+  -- That means we might override the most up-to-date value set by the other worker
+  -- with an older value in the current worker. Theoretically it's possible that
+  -- we are stuck with the same old EWMA value but for this to happen all the workers
+  -- must be sending request to the same upstream in a way that they always read old value
+  -- and override one another's EWMA value with the old one.
```

Comment:
I think this could have some other unexpected side effects, but I can't tell exactly what ATM.
Also not sure how harmful that could be.

I think, if we go with this we need to emit some statsd to have a better view of what is happening.

In parallel we can dig into the solution we were talking about on Slack, asynchronously update EWMA from worker 0 based on information send by workers using shared dict `[lr]push`.

Reply:
If you have some very strong opinion that this is safe, please move forward with this in a limited environment.